Source/Protected.swift

```diff
@@ -24,10 +24,64 @@
 
 import Foundation
 
-// MARK: -
+private protocol Lock {
+    func lock()
+    func unlock()
+}
+
+extension Lock {
+    /// Executes a closure returning a value while acquiring the lock.
+    ///
+    /// - Parameter closure: The closure to run.
+    ///
+    /// - Returns:           The value the closure generated.
+    func around<T>(_ closure: () -> T) -> T {
+        lock(); defer { unlock() }
+        return closure()
+    }
+
+    /// Execute a closure while acquiring the lock.
+    ///
+    /// - Parameter closure: The closure to run.
+    func around(_ closure: () -> Void) {
+        lock(); defer { unlock() }
+        return closure()
+    }
+}
+
+/// A `pthread_mutex_t` wrapper.
+final class MutexLock: Lock {
+    private var mutex: UnsafeMutablePointer<pthread_mutex_t>
+
+    init() {
+        mutex = .allocate(capacity: 1)
+
+        var attr = pthread_mutexattr_t()
+        pthread_mutexattr_init(&attr)
+        pthread_mutexattr_settype(&attr, .init(PTHREAD_MUTEX_ERRORCHECK))
+
+        let err = pthread_mutex_init(mutex, &attr)
+
+        precondition(err == 0, "Failed to create pthread_mutex")
+    }
+
+    deinit {
+        pthread_mutex_destroy(mutex)
+    }
+
+    fileprivate func lock() {
+        pthread_mutex_lock(mutex)
+    }
+
+    fileprivate func unlock() {
+        pthread_mutex_unlock(mutex)
+    }
+}
+
+#if !os(Linux)
```

Comment:
This might be better specified as a positive check for Apple platforms to better support arbitrary future platforms.
```suggestion
#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
```

Reply:
If we're going to do that we'd want to be specific about Linux as well.

Reply:
Updated to do exactly that. Compiling this on any other platform will fail to compile without a `lock` implemented in the `Protected` type.